DBZ-1208 Support for TLSv1.2

jordanbragg opened the pull request:
This fixed a few issues related to supporting TLSv.1.2.

### Key problem:
1. The BinaryLogClient from mysql-binlog-connector-java was used with default SSLSocketFactory settings, which ends up hard-coding the protocol to TLSv1.
2. Initial testing connection done by Debezium had issues agreeing on TLSv2 (at least in 0.8.3.Final) and required strict certificate checks.

### Fixes:
1. Did not ignore passing parameter for specifying the client side protocol (database.enableTLSProtocols=TLSv1.2)
2. Did not ignore passing parameter for specifying whether to do strict server certificate verification (database.verifyServerCertificates=false)
3. Used the property from #1 to also allow overriding the SSLSocketFactory for Shyiko's BinaryLogClient. Overridden one allows loose server certificate checking, which was the default used when using database.ssl.mode=required.

gunnarmorling commented:
Thanks a lot, @jordanbragg! We'll take a closer look asap; in the meantime, could you rework the commit so it uses an e-mail address associated to your GitHub account? That way your contribution will properly [show up](https://help.github.com/en/articles/why-are-my-contributions-not-showing-up-on-my-profile) in the stats. Thanks again!

jordanbragg commented:
Thanks @gunnarmorling for taking a look and the heads up on the commit e-mail, it should be associated now 😅 . Let me know if there is any other questions or needs.

jordanbragg commented:
In case interested:

### Verification
**Produce:**
1. Created User with REQUIRE SSL
2. tls_version variable on MySQL 5.7 DB set to "TLSv1.2"
3. Set DBZ connector to database.ssl.mode=required
4. Connector not accepted due to `Communication Link Failure - Caused by: javax.net.ssl.SSLException: Unsupported record version Unknown-0.0 `

**Fix:**
1. Added passing option for database.enableTLSProtocols=TLSv1.2
2. Connector failed for PXIX cert validation (we have self-signed certs in some environments)
3. Added passing option for database.verifyServerCertificate=false
4. Connector Accepted
5. Connector failed to start BinlogClient due to:  `Communication Link Failure - Caused by: javax.net.ssl.SSLException: Unsupported record version Unknown-0.0`
6. Added code to override BinlogClient's default SSLSocketFactory that hard codes TLSv1 to use whatever was passed into database.enableTLSProtocols variable as long as its present and SSL mode is enabled.
7. Connector Accepted and running

**Assumptions made:**
1. BinlogClient's SSLSocketFactory override would not do strict certificate validation. In the default mysql binlog library, this is the SSLSocketFactory used when ssl.mode was 'required'. If it was verify_ca, it did strict checking. I figured the initial test connection did proper certificate checking if verifyServerCertificate was not false, so it was ok to let the binlog use the "required" default.

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlJdbcContext.java:
```diff
@@ -70,6 +73,14 @@ else if ("true".equals(legacyDateTime)) {
             logger.warn("'" + JDBC_PROPERTY_LEGACY_DATETIME + "'" + " is set to 'true'. This setting is not recommended and can result in timezone issues.");
         }
 
+        if (jdbcConfig.getString(JDBC_PROPERTY_VERIFY_SERVER_CERT) != null) {
```
Is it actually needed to pass them here? They should be propagated automatically to the JDBC driver when prefixed with the "database." property prefix.

jpechane replied:
Excatly, there is an automatic passthrough so this coude should be superluous.

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:
```diff
@@ -183,6 +190,33 @@ public BinlogReader(String name, MySqlTaskContext context, HaltingPredicate acce
+
+                            @Override
+                            public void checkClientTrusted(X509Certificate[] x509Certificates, String s)
+                                throws CertificateException { }
```
I think this goes back to your last comment in the PR description, but doesn't this conflict with "strict" SSL checking?

jpechane replied:
@jordanbragg Please take a look at https://github.com/shyiko/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java#L704
You should have two factories and use theme depending on value of `sslMode` property.

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlJdbcContext.java:
```diff
@@ -37,6 +37,9 @@
 
     protected static final String MYSQL_CONNECTION_URL = "jdbc:mysql://${hostname}:${port}/?useInformationSchema=true&nullCatalogMeansCurrent=false&useSSL=${useSSL}&useUnicode=true&characterEncoding=UTF-8&characterSetResults=UTF-8&zeroDateTimeBehavior=CONVERT_TO_NULL";
     protected static final String JDBC_PROPERTY_LEGACY_DATETIME = "useLegacyDatetimeCode";
+    protected static final String JDBC_PROPERTY_VERIFY_SERVER_CERT = "verifyServerCertificate";
```
`verifyServerCertificate` is deprecedated for MySQL JDBC driver, you should look at value of `sslMode` config option.

jordanbragg commented:
@jpechane, @gunnarmorling  - Thanks for the review and sorry for any wasted time, seems I rushed a bit.  Took a step back and did more testing and cleanup. Appreciate the patience.

Latest changes removed those fields being specifically handled as they are properly coming through. Some of the confusion on my part was because whats deployed for my team now is 0.8.3.Final (latest fails for various reasons) and its mysql-connector-java library dependency is 5.1.40. enabledTLSProtocols was not accepted until 5.1.44. I also got a bit thrown off by how database.useSSL and database.sslMode are handled. Either way, bad assumptions..

The latest changes do the following:
1. Initial connection can agree based on highest common version or be forced by database.enabledTLSProtocols.
2. When setting up BinaryLogClient, query the agreed upon ssl version using SHOW SESSION STATUS LIKE 'Ssl_version'
3. Pulled over the logic from BinaryLogClient where if sslMode is Required or Preferred, use loosely checked SSLSocketFactory, if ssl is enabled and not required or preferred (verify_ca, verify_identity) then use default SSLSocketFactory (strict checking).

Let me know if there is any other concerns or anything. Again, thanks for the time and patience!

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:
```diff
@@ -1041,4 +1051,48 @@ protected BinaryLogClient getBinlogClient() {
+    protected void setBinlogSSLSocketFactory() {
+        if (connectionContext.jdbc() != null) {
+            String acceptedTLSVersion = connectionContext.getSessionVariableForSslVersion();
+            if (acceptedTLSVersion != null) {
```
Could you please use `io.debezium.util.Strings.isNullOrEmpty(String)` - better safe than sorry in case the non-value will be returned as an empty string

jordanbragg replied:
Updated, thanks!

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:
```diff
@@ -1041,4 +1051,48 @@ protected BinaryLogClient getBinlogClient() {
+                            }, null);
+                        }
+                    });
+                } else {
```
Please use
```
}
else {
```
formatting style

jordanbragg replied:
Updated, thanks!

jpechane commented:
@jordanbragg Thanks a lot for changes, it looks nice now, I like your way to get the actual TLS setting from database. I've left just two minor comments but I've one question. Right now we lack SSL test. Would it be possible to try if it is feasible now to have a test case with SSL enabled?

jordanbragg commented:
Thanks for the follow-up @jpechane. I addressed the minor comments and added a few minor test.

### Tests:
1. Verifies that setting a bad TLS version fails
2. Verifies that enforcing TLSv1.2 works and that SSL version query returns it properly

These tests verify at least that database.enabledTLSProtocols is enforced and runs and accepts TLSv1.2. In my mind, the more ideal tests would be to have mysql instances where tls_version was more restricted, so I wasn't forced to add database.enabledTLSProtocols for all tests (one my.cnf with just tls_version=TLSv1.2, for example), but it did not seem like a trivial thing. It may be a good enhancement to add a suite of tests for different versions and ssl modes.

If you feel more is needed to get a level of comfort, do let me know. Thanks again!

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:
```diff
@@ -1041,4 +1053,49 @@ protected BinaryLogClient getBinlogClient() {
     }
+
+    protected void setBinlogSSLSocketFactory() {
+        if (connectionContext.jdbc() != null) {
```
No need to check this for null.

gunnarmorling commented:
Rebased, squashed and applied. Also added a commit for doing the changes I suggested above myself. Thanks a lot, @jordanbragg!

jordanbragg commented:
Thanks @gunnarmorling ,the PR comments and changes make sense, thanks for doing that!
